File: typescript/5.最长回文子串.py
#
# @lc app=leetcode.cn id=5 lang=python3
#
# [5] 最长回文子串
#

# @lc code=start

class Solution:
    def longestPalindrome(self, s: str) -> str:
        s = "#" + "#".join(s)+"#"  # 必须两头有#否则无法处理单字符
        sLen = len(s)
        lenArr = [0]*sLen
        c, r, p = 0, 0, 0
        for i in range(sLen):
            mirror = 2*c-i  # 对称回文中心的位置
            if i < r:
                if lenArr[mirror] < r-i:
                    continue
                else:
                    lenArr[i] = r-i
            b = i+(1+lenArr[i])  # 1的作用是指向边界之外
            a = i-(1+lenArr[i])
            if a >= 0 and b < sLen and s[a] == s[b]:
                lenArr[i] += 1
                a -= 1
                b += 1
                while a >= 0 and b < sLen and s[a] == s[b]:
                    lenArr[i] += 1
                    a -= 1
                    b += 1
                c = i
                r = i+lenArr[i]
                if lenArr[p] < lenArr[i]:
                    p = i
        result = s[p-lenArr[p]+1: p+lenArr[p]]
        result = result.replace("#", "")
        return result
        # @lc code=end

File: typescript/5.最长回文子串.js
/*
 * @lc app=leetcode.cn id=5 lang=javascript
 *
 * [5] 最长回文子串
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var preprocess = function (s) {
    let str = new Array(s.length * 2 + 1)
    for (let i = 0; i < s.length; i++)
        str[2 * i] = '#', str[2 * i + 1] = s[i]
    str[str.length - 1] = '#'
    return str.join('')
}
var longestPalindrome = function (s) {
    let str = preprocess(s)
    //储存回文子串长度一半的数组
    let arr = new Array(str.length).fill(0)
    //当前回文中心
    let c = 0
    //回文子串右边界
    let r = 0
    //最长回文子串的中心位置
    let p = 0
    for (let i = 0; i < str.length; i++) {
        let mirror = 2 * c - i
        //以当前字符为回文中心的回文子串至少有这么长
        if (i < r) {
            //如果当前回文子串[严格]落在更大的回文子串内就肯定不会再扩展了
            if (arr[mirror] < r - i)
                continue
            else
                arr[i] = r - i
        }
        let a = i + (1 + arr[i])
        let b = i - (1 + arr[i])
        //扩展窗口 
        if (a < str.length && b >= 0 && str[a] == str[b]) {
            arr[i]++
            a++
            b--
            while (a < str.length && b >= 0 && str[a] == str[b]) {
                arr[i]++
                a++
                b--
            }
            //更新
            c = i;
            r = i + arr[i];
            if (arr[p] < arr[i])
                p = i
        }
    }

    //取得最终结果
    let result = str.substring(p - arr[p] + 1, p + arr[p])
    result = result.replace(/#/g, '')
    return result
};
// @lc code=end

